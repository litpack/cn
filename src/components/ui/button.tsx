import { customElement } from "solid-element";
import { mergeProps } from "solid-js";
import globalCSS from "../../index.css?inline";
import { cva, type VariantProps } from "class-variance-authority";

const buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground hover:bg-primary/90",
        destructive:
          "bg-destructive text-destructive-foreground hover:bg-destructive/90",
        outline:
          "border border-input bg-background hover:bg-accent hover:text-accent-foreground",
        secondary:
          "bg-secondary text-secondary-foreground hover:bg-secondary/80",
        ghost: "hover:bg-accent hover:text-accent-foreground",
        link: "text-primary underline-offset-4 hover:underline",
      },
      size: {
        default: "h-10 px-4 py-2",
        sm: "h-9 rounded-md px-3",
        lg: "h-11 rounded-md px-8",
        icon: "h-10 w-10",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
    },
  }
);

interface ButtonProps extends VariantProps<typeof buttonVariants> {
  disabled?: boolean;
  onClick?: (e: MouseEvent) => void;
  label?: string;
}

function Button(props: ButtonProps) {
  const mergedProps = mergeProps(
    {
      variant: "default" as VariantProps<typeof buttonVariants>["variant"],
      size: "default" as VariantProps<typeof buttonVariants>["size"],
      disabled: false,
      label: "",
    },
    props
  );

  return (
    <>
      <style>{globalCSS}</style>
      <button
        class={buttonVariants({
          variant: mergedProps.variant,
          size: mergedProps.size,
        })}
        disabled={mergedProps.disabled}
        onClick={mergedProps.onClick}
      >
        {mergedProps.label}
      </button>
    </>
  );
}

customElement(
  "cn-button",
  {
    variant: undefined,
    size: undefined,
    disabled: undefined,
    label: undefined,
    onClick: undefined,
  },
  Button
);
