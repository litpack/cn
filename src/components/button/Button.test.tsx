import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@solidjs/testing-library";
import "@testing-library/jest-dom";
import { Button } from "./button";

describe("Button Component", () => {
  it("renders with default styles and label", () => {
    render(() => <Button label="Default Button" />);
    const button = screen.getByRole("button", { name: /default button/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass("bg-primary text-primary-foreground");
  });

  it("applies the correct variant and size classes", () => {
    render(() => (
      <Button label="Destructive Button" variant="destructive" size="lg" />
    ));
    const button = screen.getByRole("button", { name: /destructive button/i });
    expect(button).toHaveClass(
      "bg-destructive text-destructive-foreground hover:bg-destructive/90"
    );
    expect(button).toHaveClass("h-11 rounded-md px-8");
  });

  
  it("handles the disabled state correctly", () => {
    render(() => <Button label="Disabled Button" disabled />);
    const button = screen.getByRole("button", { name: /disabled button/i });
    expect(button).toBeDisabled();
    expect(button).toHaveClass("disabled:pointer-events-none disabled:opacity-50");
  });

  it("triggers the onClick handler when clicked", () => {
    const onClick = vi.fn();
    render(() => <Button label="Clickable Button" onClick={onClick} />);
    const button = screen.getByRole("button", { name: /clickable button/i });
    fireEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(() => <Button label="Snapshot Button" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
