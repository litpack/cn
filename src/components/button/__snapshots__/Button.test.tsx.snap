// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > matches the snapshot 1`] = `"<style></style><button class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg]:size-4 [&amp;_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2">Snapshot Button</button>"`;

exports[`Button Component matches the snapshot 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    disabled=""
  >
    Snapshot Button
  </button>
</div>
`;
